staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,627 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fairqueuing
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/klog"
+)
+
+// QueueSetFactoryImpl implements the QueueSetFactory interface
+// QueueSetFactoryImpl makes QueueSetSystem objects.
+// This filter makes a QueueSetSystem for each priority level.
+type QueueSetFactoryImpl struct{}
+
+// NewQueueSet creates a new QueueSetSystem object
+// There is a new QueueSet created for each priority level.
+func (qsf QueueSetFactoryImpl) NewQueueSet(concurrencyLimit, desiredNumQueues, queueLengthLimit int, requestWaitLimit time.Duration, clk clock.Clock) QueueSet {
+	return NewQueueSetImpl(concurrencyLimit, desiredNumQueues, queueLengthLimit, requestWaitLimit, clk)
+}
+
+// QueueSetImpl is a fair queuing implementation designed for the kube-apiserver.
+// FQ is designed for
+// 1) dispatching requests to be served rather than packets to be transmitted
+// 2) serving multiple requests at once
+// 3) accounting for unknown and varying service times
+// implementation of:
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
+type QueueSetImpl struct {
+	Lock                 sync.Mutex
+	wg                   *sync.WaitGroup
+	Queues               []FQQueue
+	clock                clock.Clock
+	vt                   float64
+	estimatedServiceTime float64
+	lastRealTime         time.Time
+	robinIdx             int
+	// NumPacketsEnqueued is the number of packets currently enqueued
+	// (eg: incremeneted on Enqueue, decremented on Dequue)
+	NumPacketsEnqueued int
+	concurrencyLimit   int
+	desiredNumQueues   int
+	queueLengthLimit   int
+	requestWaitLimit   time.Duration
+	Quiescent          bool // emptyHandler?
```

Comment:
what's the point of making `Lock`, `Queues`, `NumPacketsEnqueued` and `Quiescent` exported/public type?

Reply:
these were heldover from when there was 2 layers, I have made them private now.